### DISCUSSION TITLE
Trino client 452 like trino jdbc should work on java 17.

### DISCUSSION THREAD

**Discussion by cloudcheflabs** [trinodb/trino#23491] [Q&A] [answered]
Hi,
I know that trino 452 which I am using can be run on java 22. 
But the problem is that trino client 452 does not work on java 17 which is our main java version on which our applications like spring boot applications are running. Trino clients like trino jdbc needs to be used in our applications running on java 17.
I think, even if trino server 452 can be run on java 22, trino clients like trino jdbc should work on another lower java versions like java 17, because not all the tools and libraries like spring boot 2 which I am using, and trino jdbc is used in are not ready to be run on java 22.

- Kidong.

**Comment by nineinchnick** [comment]
Trino JDBC driver is compatible with Java 8. Can you post details why it's not working?

**Comment by cloudcheflabs** [comment]
@nineinchnick Thanks for your reply.
You are right, I was confused with trino dependencies of trino jdbc and trino parser used in my application. Sorry for that.
trino jdbc is working on java 17.

 But I am also using trino parser in my spring boot application. trino-parser 452 seems to be compiled with java 22, so I got the following error.

```
cannot access io.trino.sql.tree.CreateTableAsSelect
[ERROR]   bad class file: xxx/io/trino/trino-parser/452/trino-parser-452.jar(/io/trino/sql/tree/CreateTableAsSelect.class)
[ERROR]     class file has wrong version 66.0, should be 61.0
```


I pulled trino source codes with 452, and I tried trino-parser with java 17 with adding the following in trino-parser subproject.
```
<build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.3</version>
                <configuration>
                    <source>17</source>
                    <target>17</target>
                    <compilerVersion>17</compilerVersion>
                    <encoding>utf-8</encoding>
                </configuration>
            </plugin>
        </plugins>
    </build>
```
I got error.

```
[ERROR] /Users/xxx/trino/core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java:[124,33] as of release 9, '_' is a keyword, and may not be used as an identifier
```

Is there any way to compile trino-parser 452 with java 17?

**Reply by nineinchnick** [reply, marked as answer]
AFAIK Trino parser is not a public API. Either update JDK in your project, or create a Trino fork where you'd apply any necessary modifications to build it with older JDK.
